kit/Kit.cpp

```diff
@@ -2365,11 +2365,40 @@ void Document::drainQueue()
         {
             std::vector<TileCombined> tileRequests = _queue->popWholeTileQueue();
 
-            // Put requests that include tiles in the visible area to the front to handle those first
-            std::partition(tileRequests.begin(), tileRequests.end(), [this](const TileCombined& req) {
-                return isTileRequestInsideVisibleArea(req); });
-            for (auto& tileCombined : tileRequests)
-                renderTiles(tileCombined);
+            const int consumedMicroS = std::chrono::duration_cast<std::chrono::microseconds>(
+                                           std::chrono::steady_clock::now() - now).count();
+            timeoutMicroS -= consumedMicroS;
+
+            if( timeoutMicroS >= consumedMicroS )
+            {
+                // Put requests that include tiles in the visible area to the front to handle those first
+                std::partition(tileRequests.begin(), tileRequests.end(), [this](const TileCombined& req) {
+                    return isTileRequestInsideVisibleArea(req); });
+                for (auto& tileCombined : tileRequests)
+                    renderTiles(tileCombined);
+
+                LOG_TRC("Kit TileCombined Items: " << tileRequests.size()
+                        << "; pre[consumed " << consumedMicroS << "us, left " << timeoutMicroS << "us]");
+            }
+            else
+            {
+                // Limited time left, push-back all invisible tile-requests
+                size_t invisibleCount = 0;
+                for(TileCombined req : tileRequests)
+                {
+                    if( isTileRequestInsideVisibleArea(req) )
+                    {
+                        renderTiles(req);
```

Comment:
Here we seem to render tiles individually one by one - which can create a massive performance problem - vs. rendering groups of them; why do we do that ?